src/plugins/vis_type_vega/public/line_vega_spec_expression.ts

```diff
@@ -0,0 +1,246 @@
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { cloneDeep } from 'lodash';
+import { i18n } from '@osd/i18n';
+import {
+  ExpressionFunctionDefinition,
+  OpenSearchDashboardsDatatable,
+} from '../../expressions/public';
+import { VegaVisualizationDependencies } from './plugin';
+
+type Input = OpenSearchDashboardsDatatable;
+type Output = Promise<string>;
+
+interface Arguments {
+  visLayers: string | null;
+  visParams: string;
+  dimensions: string;
+}
+
+export type VegaSpecExpressionFunctionDefinition = ExpressionFunctionDefinition<
+  'vega_spec',
+  Input,
+  Arguments,
+  Output
+>;
+
+const createSpecFromDatatable = (
+  datatable: OpenSearchDashboardsDatatable,
+  visParams: string,
+  dimensionsString: string
+): object => {
+  // TODO: we can try to use VegaSpec type but it is currently very outdated, where many
+  // of the fields and sub-fields don't have other optional params that we want for customizing.
+  // For now, we make this more loosely-typed by just specifying it as a generic object.
+  const spec = {} as any;
+  const xAxis = datatable.columns[0];
+
+  const parseParams = JSON.parse(visParams);
+  const dimensions = JSON.parse(dimensionsString);
+  const legendPosition = parseParams.legendPosition;
+
+  // Get time range for the data in case there is only data for a small range so it will show the full time range
+  const startTime = {};
+  const xAxisId = xAxis.id.toString();
+  // @ts-ignore
+  startTime[xAxisId] = new Date(dimensions.x.params.bounds.min).valueOf();
+  const endTime = {};
+  // @ts-ignore
+  endTime[xAxisId] = new Date(dimensions.x.params.bounds.max).valueOf();
+  const updatedTable = datatable.rows.concat([startTime, endTime]);
+
+  // TODO: update this to v5 when available
+  spec.$schema = 'https://vega.github.io/schema/vega-lite/v4.json';
+  spec.data = {
+    values: updatedTable,
+  };
+  spec.config = {
+    view: {
+      stroke: null,
+    },
+    concat: {
+      spacing: 0,
+    },
+    // the circle timeline representing annotations
+    circle: {
+      color: 'blue',
+    },
+    // the vertical line when user hovers over an annotation circle
+    rule: {
+      color: 'red',
+    },
+    legend: {
+      orient: legendPosition,
+    },
+  };
+
+  // Get the valueAxes data and generate a map to easily fetch the different valueAxes data
+  const valueAxis = {};
+  parseParams.valueAxes.forEach((yAxis: { id: { toString: () => string | number } }) => {
```

Comment:
If there are no value axes, will this throw an NPE? Or is `valueAxes` defaulted to an empty array if there are none?

Reply:
Fixing this to be more robust. It should be empty an array.